indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisorStateManager.java

```diff
@@ -0,0 +1,334 @@
+
+  private State supervisorState;
+  private final int healthinessThreshold;
+  private final int unhealthinessThreshold;
+  private final int healthinessTaskThreshold;
+  private final int unhealthinessTaskThreshold;
+
+  private boolean atLeastOneSuccessfulRun = false;
+  private boolean currentRunSuccessful = true;
+  private final CircularBuffer<TaskState> completedTaskHistory;
+  private final CircularBuffer<State> stateHistory; // From previous runs
+  private final ExceptionEventStore eventStore;
+
+  public SeekableStreamSupervisorStateManager(
+      State initialState,
+      SeekableStreamSupervisorConfig supervisorConfig
+  )
+  {
+    this.supervisorState = initialState;
+    this.healthinessThreshold = supervisorConfig.getHealthinessThreshold();
+    this.unhealthinessThreshold = supervisorConfig.getUnhealthinessThreshold();
+    this.healthinessTaskThreshold = supervisorConfig.getTaskHealthinessThreshold();
+    this.unhealthinessTaskThreshold = supervisorConfig.getTaskUnhealthinessThreshold();
+    Preconditions.checkArgument(
+        supervisorConfig.getMaxStoredExceptionEvents() >=
+        Math.max(healthinessThreshold, unhealthinessThreshold),
+        "numExceptionEventsToStore must be greater than or equal to both "
+        + "healthinessThreshold and unhealthinessThreshold"
+    );
+    this.eventStore = new ExceptionEventStore(
+        supervisorConfig.getMaxStoredExceptionEvents(),
+        supervisorConfig.isStoringStackTraces()
+    );
+    this.completedTaskHistory = new CircularBuffer<>(Math.max(healthinessTaskThreshold, unhealthinessTaskThreshold));
+    this.stateHistory = new CircularBuffer<>(Math.max(healthinessThreshold, unhealthinessThreshold));
+  }
+
+  public Optional<State> setStateIfNoSuccessfulRunYet(State newState)
+  {
+    if (!atLeastOneSuccessfulRun) {
+      return Optional.of(setState(newState));
+    }
+    return Optional.absent();
+  }
+
+  public State setState(State newState)
+  {
+    if (newState.equals(State.SUSPENDED)) {
+      atLeastOneSuccessfulRun = false; // We want the startup states again after being suspended
+    }
+    this.supervisorState = newState;
+    return newState;
+  }
+
+  public void storeThrowableEvent(Throwable t)
+  {
+    if (t instanceof PossiblyTransientStreamException && atLeastOneSuccessfulRun) {
+      t = new TransientStreamException(t);
+    }
+
+    eventStore.storeThrowable(t);
+    currentRunSuccessful = false;
+  }
+
+  public void storeCompletedTaskState(TaskState state)
+  {
+    completedTaskHistory.add(state);
+  }
+
+  public void markRunFinishedAndEvaluateHealth()
+  {
+    if (currentRunSuccessful) {
+      atLeastOneSuccessfulRun = true;
+    }
+
+    State currentRunState = State.RUNNING;
+
+    for (Map.Entry<Class, Queue<ExceptionEvent>> events : eventStore.getNonTransientRecentEvents().entrySet()) {
+      if (events.getValue().size() >= unhealthinessThreshold) {
+        if (events.getKey().equals(NonTransientStreamException.class)) {
```

Comment:
Is it possible that you once connected successfully to the stream, but later get a lot of `NonTransientStreamException`?

Reply:
Yes.  Although it shouldn't happen in normal operation, I could imagine that there's some edge cases where that might occur (e.g. someone changing a Kafka cluster's permissions causing an already successfully running supervisor to start throwing auth exceptions?).   Do you think there's value in subdividing the `LOST_CONTACT_WITH_STREAM` state into `LOST_CONTACT_WITH_STREAM_NON_TRANSIENT` and `LOST_CONTACT_WITH_STREAM_TRANSIENT` states?

Reply:
Hm, I'm not sure that tying transience to the type of exception is the right approach (I think transience is more about the number of consecutive failures regardless of the exception type).

Since the error states are trying to convey whether the supervisor is having stream connectivity issues (UNABLE_TO_CONNECT_TO_STREAM, LOST_CONTACT_WITH_STREAM) or if its some other kind of issue (UNHELAHTY_SUPERVISOR), I'm thinking it would be better to separate the exceptions into two categories:
- Stream connection problems
- problems unrelated to stream connectivity

UNABLE_TO_CONNECT_TO_STREAM would be the state when you have stream connection exceptions over the configured threshold, and you've not successfully connected before

LOST_CONTACT_WITH_STREAM would occur when you have stream connection exceptions over the configured threshold, and you have successfully connected before

UNHEALTHY_SUPERVISOR would occur when you have non-stream connection exceptions over the threshold

Reply:
Although I agree that your implementation would result in a cleaner API design, I still think the transience of an error is important to convey since it indicates severity.  For example, a Kinesis supervisor that encounters `LimitExceededException`s (a temporary API limit exception) over the first `threshold` runs shouldn't be treated equally in severity to a supervisor that throws some sort of auth exception over the first `threshold runs` since only the former could possibly recover without user intervention.  If we were to eliminate that concept, there'd be no difference from the API caller's perspective.

In the initial discussion on this change's proposal, @dclim said "If it transitions from RUNNING (healthy) to UNHEALTHY, assume that someone has hooked up a monitoring system to it and is going to get paged at 4am in the morning, so it better actually be UNHEALTHY, and not some transient error that is going to resolve in the next minute".  If we're set on working under the assumption that someone's hooking up this endpoint to a pager, I think the transience of the error is very relevant.

Reply:
> If we're set on working under the assumption that someone's hooking up this endpoint to a pager, I think the transience of the error is very relevant.

If the exceptions have different severities, then I think you could be more lenient on the failure thresholds for low severity errors. UNABLE_TO_CONNECT and LOST_CONTACT should accurately reflect the real state/history though, I don't think fudging that for low severity errors is the right approach.

Reply:
Another consideration might be to have the status report contain some error severity suggestion, if you only had low severity errors the report could indicate that somehow and a user that didn't care too much low sev errors could choose not to alert in such situations, also, is the classification of low sev vs high sev errors very clear?

Reply:
Splitting the states into transient vs non transient could be useful, but naybe better to indicate transience or severity in a separate field?

Reply:
I can add a field to `ExceptionEvent` that indicates transience then have the states set regardless of error transience.  Is that approach cool?

Reply:
@justinborromeo  That sounds good to me